File: SOLID/LSP/Compliant/Main2.java
abstract class Bird{
    abstract void eat();
}

interface Flyable{
    void fly();
}
interface Swimmable{
    void swim();
}

class Sparrow extends Bird implements Flyable{
    public void eat(){
        System.out.println("Sparrow is eating");
    }
    public void fly(){
        System.out.println("Sparrow is flying");
    }
}

class Penguin extends Bird implements Swimmable{
    public void eat(){
         System.out.println("Penguin is eating");
    }
    public void swim(){
        System.out.println("Penguin is swimming");
    }
}
public class Main2{
    public static void main(String[] args) {

        Bird penguin = new Penguin();
        Bird sparrow = new Sparrow();

        penguin.eat(); 
        sparrow.eat();  

        Flyable flyingBird = new Sparrow();
        flyingBird.fly(); 

        Swimmable penguinbird = new Penguin();

        penguinbird.swim();
    }
}

File: SOLID/LSP/Compliant/Main2.ts
// Abstract class
abstract class Bird {
  abstract eat(): void;
}

// Interfaces
interface Flyable {
  fly(): void;
}

interface Swimmable {
  swim(): void;
}

// Sparrow class
class Sparrow extends Bird implements Flyable {
  eat(): void {
    console.log("Sparrow is eating");
  }

  fly(): void {
    console.log("Sparrow is flying");
  }
}

// Penguin class
class Penguin extends Bird implements Swimmable {
  eat(): void {
    console.log("Penguin is eating");
  }

  swim(): void {
    console.log("Penguin is swimming");
  }
}

// Main
const penguin: Bird = new Penguin();
const sparrow: Bird = new Sparrow();

penguin.eat();
sparrow.eat();

const flyingBird: Flyable = new Sparrow();
flyingBird.fly();

const penguinBird: Swimmable = new Penguin();
penguinBird.swim();
